### DISCUSSION TITLE
Add Compiler Generator through a Compiler Generator Language

### DISCUSSION THREAD

**Discussion by sirinath** [jzimmerman/langcc#29] [Ideas]
Can backend generation also be applied to the project where there can be compiler language which transforms or rewrites a AST to machine or byte/intermediate code  depending on the target.

**Comment by jzimmerman** [comment]
Unfortunately, this is not so straightforward. Part of the appeal of _frontend_ generation is that it can be done programmatically, from a simple declarative BNF description (see https://arxiv.org/pdf/2209.08383.pdf for details). But what would be the BNF-like language from which one would generate a backend? Most of the time, a full "declarative backend spec" is simply a functional program that translates source AST to target AST - and this is just what a backend is.

**Reply by sirinath** [reply]
It can be a language with term/tree rewriting or reduction system which takes the source AST and transforms it to the the target AST.

**Reply by sirinath** [reply]
Compiling and optimising can be specified in the same way in a declarative way.

**Reply by jzimmerman** [reply]
Can you write an example of such a declarative backend spec? (E.g., from a toy Python-like language to a toy assembly language?) I suspect that if you want to produce a real optimizing compiler, the declarative backend spec will look a lot like general-purpose code.

**Reply by sirinath** [reply]
Something like given there: 
- [Tree Rewriting](https://program-transformation.org/Transform/TreeRewriting)

More other details are at [Program transformation](https://program-transformation.org/Transform/ProgramTransformation.html), [Program transformation](https://program-transformation.org/Transform/WebHome.html)

**Reply by jzimmerman** [reply]
I think that one might be able to use tree-rewriting systems for some very simple transformations (roughly, tree homomorphisms), but not for a full optimizing compiler. Full compiler backends of any reasonable scale tend to require general-purpose programming features.

**Reply by sirinath** [reply]
See this: https://egraphs-good.github.io. This uses high level language bindings but I feel it might be possible to distill it down to a declarative language.

**Reply by jzimmerman** [reply]
Maybe, but I think that would only represent a relatively small subset of the transformations one would like to do in a full compiler backend.

I'd be happy to be proven wrong -- if you have an example of what a full compiler backend would look like in some simple declarative DSL (e.g., for the Python-like toy language I mentioned above), I'd be interested to see it.

**Reply by sirinath** [reply]
This is not exactly this but related:
- https://github.com/Maratyszcza/PeachPy
- https://github.com/Maratyszcza/Opcodes
- https://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.152.5846&rep=rep1&type=pdf
- https://www.cs.tufts.edu/~nr/pubs/specifying.pdf
- https://www.cs.tufts.edu/~nr/pubs/gentileset.pdf
- https://cs.stackexchange.com/questions/16058/compiler-backend-generated-from-declarative-description-of-the-architecture

**Reply by sirinath** [reply]
Also there is: https://kframework.org

**Reply by jzimmerman** [reply]
Yes, I think significant parts of compiler backend development can be automated, but often each in different ways -- and not to the extent that one can generate a complete backend automatically. As such, I believe it's still out of scope for a tool like langcc.
